src/group_theory/free_abelian_group.lean

```diff
@@ -345,7 +398,68 @@ instance [comm_monoid α] : comm_ring (free_abelian_group α) :=
       { intros y1 y2 ih1 ih2, rw [mul_add, add_mul, ih1, ih2] } },
     { intros s ih, rw [neg_mul_eq_neg_mul_symm, ih, neg_mul_eq_mul_neg] },
     { intros x1 x2 ih1 ih2, rw [add_mul, mul_add, ih1, ih2] }
-  end
+  end,
   .. free_abelian_group.ring α }
 
+/-- The free abelian group on the empty type is the trivial group. -/
+def pempty_equiv : free_abelian_group pempty ≃+ unit :=
+{ to_fun := lift (λ _, ()),
+  inv_fun := λ _, 0,
+  left_inv := λ z, free_abelian_group.induction_on z
+    ( rfl )
+    ( λ x, pempty.elim x)
+    ( λ x, pempty.elim x)
+    ( λ x y hx hy, by {simp only * at *, simp * at *}),
+  right_inv := λ z, unit.ext,
+  map_add' := λ x y, unit.ext }
+
+/-- The free abelian group on `punit` is isomorphic to `ℤ`. -/
+def punit_equiv : free_abelian_group punit ≃+ ℤ :=
+{ to_fun := free_abelian_group.lift (λ _, (1 : ℤ)),
+  inv_fun := λ n, n • of (punit.star),
+  left_inv := λ z, free_abelian_group.induction_on z
+    (by {dsimp only, rw [add_monoid_hom.map_zero, zero_smul]})
+    (λ x, punit.cases_on x (by simp))
+    (λ x, punit.cases_on x (by simp))
+    (λ x y hx hy, by { simp only [lift.add, add_smul] at *, rw [hx, hy]}),
+  right_inv := λ n,
+  begin
+    rw [add_monoid_hom.map_int_module_smul, lift.of],
+    exact gsmul_int_one n
+  end,
+  map_add' := add_monoid_hom.map_add _ }
+
+-- this would be easier with `to_fun := map` but for some reason `map` is not universe
+-- polymorphic.
```

Comment:
It would be perfectly possible to make `map` universe polymorphic, I guess. It's done the way it is now to fit nicely with the `monad free_abelian_group` instance.

It doesn't seem an exciting change, however, so unless someone wants to otherwise, I don't think it's necessary for this PR.

Reply:
I had assumed that somehow `map` was too buried in the monadology to extract, but actually it's not. I will fix this.

Reply:
Done!